File: HashMap/kDiffPairs.cpp
class Solution {
public:
    int findPairs(vector<int>& nums, int k) {
        int numPairs = 0;
        
        set<int> targets;
        set<vector<int>> pairsSeen;
        
        int n = nums.size();
        for (int i = 0; i < n; ++i){
            
            vector<int> hi{nums[i], nums[i] + k};
            vector<int> lo{nums[i] - k, nums[i]};
            
            if (targets.find(hi[1]) != targets.end() && pairsSeen.find(hi) == pairsSeen.end()){
                ++numPairs;
                pairsSeen.insert(hi);
            }
            if (targets.find(lo[0]) != targets.end() && pairsSeen.find(lo) == pairsSeen.end()){
                ++numPairs;
                pairsSeen.insert(lo);
            }
            targets.insert(nums[i]);
        }
        
        return numPairs;
    }
};

File: Maths Problems/kDiffPairs.py
# https://leetcode.com/problems/k-diff-pairs-in-an-array/

# Version 1 - Counters hashmap
# Time complexity       O(n)
# Space complexity      O(n)
from collections import Counter
class Solution:
    def findPairs(self, nums: List[int], k: int) -> int:
        countedNums = Counter(nums)
        numPairs = 0
        for num in countedNums:
            if num + k in countedNums and k > 0:
                numPairs += 1
            elif k == 0 and countedNums[num] > 1:
                numPairs += 1
        return numPairs